### ISSUE TITLE
improve halt(8)

### ISSUE THREAD

**Issue by archenemies** [systemd/systemd#13254]
Just a comment on the documentation, looking at the manual page that comes up when I do `man reboot`:

1. At the bottom it says

> These are legacy commands available for compatibility only.

If they are "legacy commands", I would explain what commands should be used as replacements, maybe:

> These are legacy commands available for compatibility only. You can also use 'systemctl halt', 'systemctl poweroff', or 'systemctl reboot'.

2. At the top,

> halt, poweroff, reboot may be used to halt, power-off or reboot the machine.

This is a bit hard to parse, I would say something like

> 'halt', 'poweroff', and 'reboot' may be used to halt, power-off or reboot the machine, respectively. All three commands take the same options.

3. At the very bottom

> These commands are implemented in a way that preserves compatibility with the original SysV commands.

Is this true? I seem to recall 'halt' being equivalent to 'poweroff'. Although looking [here](https://manpages.debian.org/unstable/sysvinit-core/halt.8.en.html), there do seem to be two separate commands for "halt" and "poweroff". And it says 2001 at the bottom...

**Comment by keszybz** [comment]
> These are legacy commands available for compatibility only.

This sentence should be removed. Those commands are never going to be removed, they are supported and in fact primary interfaces.

> Is this true? I seem to recall 'halt' being equivalent to 'poweroff'. 

No, halt "halts", i.e. leaves the machine powered-on but without the OS, and poweroff does the thing we normally want.

Any chance you could submit this as PR?

**Comment by floppym** [comment]
> Is this true? I seem to recall 'halt' being equivalent to 'poweroff'. Although looking here, there do seem to be two separate commands for "halt" and "poweroff". And it says 2001 at the bottom...

On most sysvinit-based systems, calling `halt` or `poweroff` would end up powering off the machine. That's because of a special behavior documented in the `halt(8)` man page from sysvinit:

> If halt or reboot is called when the system is not in runlevel 0 or 6, in other words when it's running normally, shutdown will be invoked instead (with the -h or -r flag). For more info see the shutdown(8) manpage.

The shutdown command would then call `telinit 0` to switch the system to runlevel 0. The last command in runlevel 0 is typically something like this:
```
l0s:0:wait:/sbin/halt -dhnp
```
And the -p flag causes the system to be powered off.

systemd does a better job of differentiating the two actions by defining two targets (halt.target and poweroff.target) instead of a single "0" runlevel like sysvinit. That means it can carry the correct verb all the way thorough to the final `reboot(2)` system call in the systemd-shutdown command.

**Comment by archenemies** [comment]
That's very useful data @floppym. This whole time I'd been thinking that Systemd was just being needlessly reinventive.

It would be good if it could go in the manual page under `NOTES`.

I can submit a PR if you like, or you can do one with your preferred wording. Just let me know.

**Comment by floppym** [comment]
I would consider the sysvinit behavior to be a bug really: `halt` ends up doing the wrong thing.

I'm not sure the systemd manual page is the right place to document a bug in sysvinit.

**Comment by archenemies** [comment]
>I'm not sure the systemd manual page is the right place to document a bug in sysvinit.

It sounds like it was a limitation, not a bug. And a lot of people are probably in the same position as myself, wondering why 'halt' does something different under Systemd. The manual page should either explain the reason for the change of behavior, or link to a document which does. Since the manual page mentions compatibility, I think a short explanation under that heading would be proper.

**Comment by floppym** [comment]
Ok, I have no strong preference, and I don't make the decisions anyway. ;-)

**Comment by archenemies** [comment]
OK well I guess I'll work on a PR then, taking your concern as much as possible into account.

**Comment by archenemies** [comment]
PR #13264 created.

**Comment by archenemies** [comment]
@keszybz

    <para>Note that on many SysV systems <command>halt</command> used to be synonymous to
    <command>poweroff</command>, i.e. both commands would equally result in powering the machine off. systemd
    is more accurate here, and <command>halt</command> results in halting the machine only (leaving power
    on), and <command>poweroff</command> is required to actually power it off.</para>

Thanks, but I still think you're underselling Systemd a bit. If my PR was too detailed, we could at least say something like:

    <para>Note that on many SysV systems <command>halt</command> used to be synonymous to
    <command>poweroff</command>, i.e. both commands would equally result in powering the machine off. This convergence was not a design feature, but was due to practical limitations of SysV, in that both verbs were implemented by switching to the same "runlevel zero". Systemd
    is able to be more accurate here, and <command>halt</command> results in halting the machine only (leaving power
    on), while <command>poweroff</command> is required to actually power it off.</para>
